libs/mrwebrtc-unityplugin/src/native_renderer.cpp

```diff
@@ -133,45 +167,74 @@ void NativeRenderer::DisableRemoteVideo() {
 }
 
 void NativeRenderer::I420ARemoteVideoFrameCallback(
-    void* user_data,
-    const mrsI420AVideoFrame& frame) {
-
-    NativeRenderer* nativeVideo = static_cast<NativeRenderer*>(user_data);
+  void* user_data,
+  const mrsI420AVideoFrame& frame) {
 
-    // RESEARCH: Do we need to keep a frame queue or is it fine to just render
-    // the most recent frame?
+  // It is possible for one buffer to be empty, each buffer must be checked.
+  if (frame.ydata_ == nullptr || frame.udata_ == nullptr || frame.vdata_ == nullptr) {
+    return;
+  }
 
-    // The performance trade-off being made here is to lock g_lock two times,
-    // preferring to copy the frame buffer outside any lock. Alternatively, the
-    // copy operation could be done inside g_lock scope. This would result in a
-    // single g_lock, but the lock would be held for a longer period of time
-    // every video frame, for every video stream. RESEARCH: Which is better?
+  NativeRenderer* nativeVideo = static_cast<NativeRenderer*>(user_data);
+ 
+  {
+    // Instance lock in case they get updated from UpdateTextures
+    std::lock_guard guard(nativeVideo->m_lock);
+    if ((int)frame.width_ != nativeVideo->m_remoteTextures[0].width || (int)frame.height_ != nativeVideo->m_remoteTextures[0].height) {
+      if (g_textureSizeChangeCallback != nullptr) {
+        g_textureSizeChangeCallback(frame.width_, frame.height_, nativeVideo->m_handle);
+      }
+      return;
+    }
+  }
 
-    // Acquire a video frame buffer from the free list.
-    {
-        // Global lock
-        std::lock_guard guard(g_lock);
-        if (nativeVideo->m_nextI420RemoteVideoFrame == nullptr) {
-            // TODO: Encapsulate this logic.
-            if (g_freeI420VideoFrames.size()) {
-                nativeVideo->m_nextI420RemoteVideoFrame = g_freeI420VideoFrames.back();
-                g_freeI420VideoFrames.pop_back();
-            } else {
-                nativeVideo->m_nextI420RemoteVideoFrame = std::make_shared<I420VideoFrame>();
-            }
-        }
+  // Acquire a video frame buffer from the free list with the global lock.
+  std::shared_ptr<I420VideoFrame> remoteI420Frame = nullptr;
+  {
+    // Global lock
+    std::lock_guard guard(g_lock);
+    if (g_freeI420VideoFrames.size()) {
+        remoteI420Frame = std::move(g_freeI420VideoFrames.back());
+        g_freeI420VideoFrames.pop_back();
+    } else {
+      try {
```

Comment:
nit: move that block outside the lock, there's no need for holding it while doing some expensive allocation.